pkg/daemon/ceph/osd/volume.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2018 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package osd
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+	"strconv"
+	"syscall"
+
+	"github.com/rook/rook/pkg/clusterd"
+	oposd "github.com/rook/rook/pkg/operator/ceph/cluster/osd"
+	"github.com/rook/rook/pkg/util/exec"
+)
+
+func (a *OsdAgent) configureDevices(context *clusterd.Context, devices *DeviceOsdMapping) ([]oposd.OSDInfo, bool, error) {
+	var osds []oposd.OSDInfo
+
+	useCephVolume, err := getCephVolumeSupported(context)
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to detect if ceph-volume is available. %+v", err)
+	}
+	if !useCephVolume {
+		return osds, false, nil
+	}
+
+	if devices == nil || len(devices.Entries) == 0 {
+		logger.Infof("no new devices to configure. returning devices already configured with ceph-volume.")
+		osds, err = getCephVolumeOSDs(context, a.cluster.Name)
+		if err != nil {
+			logger.Infof("failed to get devices already provisioned by ceph-volume. %+v", err)
+		}
+		return osds, true, nil
+	}
+
+	err = createOSDBootstrapKeyring(context, a.cluster.Name)
+	if err != nil {
+		return nil, true, fmt.Errorf("failed to generate osd keyring. %+v", err)
+	}
+
+	volumeArgs := []string{"lvm", "batch", "--prepare", "--bluestore", "--yes"}
```

Comment:
Any particular reason you're using `lvm batch --prepare disk1 disk2 ... diskN` instead of `lvm prepare disk1`, `lvm prepare disk2` ... `lvm prepare diskN`? `lvm prepare` doesn't need special nautilus support, and also supports `--block.db` and `--block.wal` (for possible future improvements), which `lvm batch` doesn't.

Reply:
@lbogdan Good questions... The idea with `lvm batch` is that it can make smarter decisions about using an ssd/nvme as the device for the db and wal, while using hdds as the data devices. The ceph-volume changes are being backported to luminous and mimic, so the new flag isn't expected to be a problem. If there is some option `batch` doesn't yet support, we should highlight it. We want ceph-volume to keep as much of the business logic around provisioning osds as possible.

Reply:
I think the question here is: should you delegate those "smarter decisions" to `ceph-volume batch`? For example the rules for [bluestore](http://docs.ceph.com/docs/mimic/ceph-volume/lvm/batch/#bluestore) look quite arbitrary to me - I, as a user, would certainly not expect that behavior.

Reply:
I suggest we explicitly call out the "fast" devices for db/wal, so the user has to specify they be used that way.  By default, if there is a mix of devices, if you just feed them to rook they'd end up as different OSDs.  This is what the Ceph orchestrator_cli interface does, so making this match should make things easier.  @sebastian-philipp

Reply:
@liewegas What you're saying is that we shouldn't use batch mode unless the fast devices are specified, right?

Rook already has a `metadataDevice` [setting](https://rook.io/docs/rook/master/ceph-cluster-crd.html#osd-configuration-settings) in the cluster.yaml that would fit the purpose of specifying fast devices. In that case, rook would use batch mode and pass the metadatadevice and hdds in batch mode.

Reply:
I mean I thought we changed (or were changing) the batch mode so that you explicitly listed which devices were the fast ones, and only then would batch mode use them for db or journal or whatever ceph-volume thinks is best.  Otherwise, batch is just doing N separate OSDs.  I may have lost track of what was implemented vs backported, though!  @alfredodeza ?

Reply:
I too think that an opt-in for the "smart defaults" (which could be enabled by specifying fast devices in `metadataDevice`, for example) is preferred, as having those "smart defaults" be the default is not something that end users are expecting.

Reply:
@liewegas @alfredodeza I don't see the option in the latest master build to specify the fast devices. Unless I'm missing the option and given the need to get this in, for now if the metadata device is specified i'll fall back to rook's implementation of the osds.

Reply:
@liewegas @travisn @lbogdan I'm still working on the extension to batch (where one can explcitily specfiy data devices, wal devices and db devices). My apologies that its taking a bit longer then expected. The tracker issue (http://tracker.ceph.com/issues/37086) has a link to the current dev branch.

Reply:
Okay, in that case let's just make sure the CRD arguments line up with the soon-to-arrive ceph-volume capability, but for the current release just feed it to the existing batch command.

Reply:
(i.e., no new independentDevices: "false" option)

Reply:
@jan--f is `--wal-devices` the main option we are referring to? This will map well to the existing `metadataDevice` setting in the CRD.

Reply:
@travisn My intention is to have --wal-devices and --db-devices for bluestore and --journal-devices for filestore. both objectstores understand --data-devices of course.

The --db-devices is just in place since bluestore can techincally be deployed on 3 devices, spinner, ssd and nvme for example. This is mostly there for completeness sake, I expect most deployment will ignore it.

To answer your question: yes I think rooks metadataDevices maps well --wal-devices (or --journal-devices for the filestore case).

Reply:
s/wal/db/.  wal should only be used when there are 3 types devices; when there are only two use --db-devices.

Reply:
@liewegas oh really? I had a chat with @ifed01 recently and took away that db is the one potentially going on very fast devices. Will double check...